core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1099,22 +1099,18 @@ public String authority() {
 
       /** Called when it's ready to create a real call and reprocess the pending call. */
       void reprocess() {
-        getCallExecutor(callOptions).execute(
-            new Runnable() {
-              @Override
-              public void run() {
-                ClientCall<ReqT, RespT> realCall;
-                Context previous = context.attach();
-                try {
-                  realCall = newClientCall(method, callOptions);
-                } finally {
-                  context.detach(previous);
-                }
-                setCall(realCall);
-                syncContext.execute(new PendingCallRemoval());
-              }
-            }
-        );
+        ClientCall<ReqT, RespT> realCall;
+        Context previous = context.attach();
+        try {
+          realCall = newClientCall(method, callOptions);
+        } finally {
+          context.detach(previous);
+        }
+        Runnable toRun = setCall(realCall);
+        if (toRun != null) {
+          getCallExecutor(callOptions).execute(toRun);
+        }
+        syncContext.execute(new PendingCallRemoval());
```

Comment:
As I said before, if `toRun != null` this will need to be delayed until `toRun` finishes executing.

Reply:
well, how can we guarantee `toRun` finishes earlier than this? 
I think it is safe to remove immediately from pendingCalls list, no?

Reply:
Oh, is that why you are making sure start() isn't part of `toRun`? As otherwise `pendingCalls` won't keep the channel alive properly during a shutdown or the like?

I thought I had mentioned this:
```java
if (toRun == null) {
  syncContext.execute(new PendingCallRemoval());
} else {
  getCallExecutor(callOptions).execute(new Runnable() {
    @Override public void run() {
      toRun.run();
      syncContext.execute(new PendingCallRemoval());
    }
  });
}